shared/cryptography/codeql/cryptography/Model.qll

```diff
@@ -22,7 +22,56 @@ module CryptographyBase<LocationSig Location, InputSig<Location> Input> {
     UnknownPropertyValue() { this = "<unknown>" }
   }
 
-  abstract class NodeBase instanceof LocatableElement {
+  private string getPropertyAsGraphString(NodeBase node, string key) {
+    result =
+      strictconcat(any(string value, Location location, string parsed |
+            node.properties(key, value, location) and
+            parsed = "(" + value + "," + location.toString() + ")"
+          |
+            parsed
+          ), ","
+      )
+  }
+
+  predicate nodes_graph_impl(NodeBase node, string key, string value) {
+    key = "semmle.label" and
+    value = node.toString()
+    or
+    // CodeQL's DGML output does not include a location
+    key = "Location" and
+    value = node.getLocation().toString()
+    or
+    // Known unknown edges should be reported as properties rather than edges
+    node = node.getChild(key) and
+    value = "<unknown>"
+    or
+    // Report properties
+    value = getPropertyAsGraphString(node, key)
+  }
+
+  predicate edges_graph_impl(NodeBase source, NodeBase target, string key, string value) {
+    key = "semmle.label" and
+    target = source.getChild(value) and
+    // Known unknowns are reported as properties rather than edges
+    not source = target
+  }
+
+  newtype TNode =
+    THashOperation(LocatableElement e) or
+    THashAlgorithm(LocatableElement e) or
+    TKeyDerivationOperation(LocatableElement e) or
+    TKeyDerivationAlgorithm(LocatableElement e) or
+    TEncryptionOperation(LocatableElement e) or
+    TSymmetricAlgorithm(LocatableElement e) or
+    TEllipticCurveAlgorithm(LocatableElement e) or
```

Comment:
This `newtype` currently generates a table of size 8 * "number of `LocatableElement`s", and I doubt that many of those rows are actually necessary.

An alternative could be something like:
```ql
abstract class EllipticCurveAlgorithmImpl instanceof LocatableElement {
  string toString() { result = super.toString() }
}

newtype TNode =
// ...
TEllipticCurveAlgorithm(EllipticCurveAlgorithmImpl e) or
// ...
```
and it's then possible to extend the set of elliptic curve algorithms by extending the `EllipticCurveAlgorithmImpl` class.

As a bonus, it's now possible to require additional stuff on e.g., elliptic curves by adding abstract predicates to `EllipticCurveAlgorithmImpl`.